pennylane_rigetti/qc.py

```diff
@@ -170,20 +170,14 @@ def _get_timeout_args(self, **kwargs) -> Dict[str, float]:
         return timeout_args
 
     def define_wire_map(self, wires):
-        if hasattr(self, "wiring"):
-            device_wires = Wires(self.wiring)
-        else:
-            # if no wiring given, use consecutive wire labels
-            device_wires = Wires(range(self.num_wires))
-
+        device_wires = Wires(self.wiring)
```

Comment:
The current implementation leads to this behaviour, which I believe is incorrect:

![image](https://github.com/PennyLaneAI/pennylane-rigetti/assets/38584660/d03437c7-0437-4ecc-826f-2a41e48c65ef)

With the above suggested change I instead get:

![image](https://github.com/PennyLaneAI/pennylane-rigetti/assets/38584660/6230bf0f-c89d-41d8-9f9a-b364a0d30488)

Does that look like the expected program for this circuit? Or have I misunderstood the program format for gates?

Reply:
Ah, okay, I misunderstood how you want the program to look. I assumed that the qubit specification for the RX operator would be the index, not the qubit number. In that case, the issue is that we are trying to use this function for two different and incompatible things.

The version I suggested in my initial comment will work for the part of PennyLane where you are encountering an issue, but will not work for generating the program for Rigetti. You've updated it to work for the Rigetti program, but not for PennyLane post processing 😅 

There are two separate types of wire mapping we are interested in here. One is the conversion between the wires on the (software) device and consecutive indexing in PennyLane. The other is the conversion between the wires on the (software) device and the qubit numbers expected by the hardware backend. This function used (and is expected by the PennyLane device interface) to do the first thing, and you've updated it to do the second.

In other words, the dictionary generated by this function IS (or at least should be) the reverse lookup table for mapping wire label to relative position in the sample array. It should not include the qubit numbers at all, but only the (software) device wire labels and the consecutive integers. I.e. a 3 wire device with `wires=["a", "b", "c"]`, this function should return `{"a": 0, "b": 1, "c":2}`. 

I don't think the plugin actually needs a custom implementation of this method, it can inherit from the PennyLane device base class. Where it does need a custom implementation is for mapping to hardware instructions, that will need to be done in a separate function, instead of overwriting this existing PennyLane function.

I'm going to make a branch of off this with a proposed solution. It seems like this same assumption was made in the parent class, i.e. that "custom wire labels" (this is what we call it whenever users don't have consecutive integers starting with 0 as their wire labels) in PennyLane don't currently translate to the Rigetti plugin. 

Let's see if my suggestion fixes this, and also if it breaks other things, and then take it from there!

Reply:
Are you sure you want the operators to be specified by qubit label and not by index? That's not currently the behaviour on `master` from what I can see.